tiledb/sm/array_schema/dimension.cc

```diff
@@ -733,35 +733,87 @@ double Dimension::overlap_ratio<char>(const Range& r1, const Range& r2) {
+  assert(r1_low <= r1_high);
+  assert(r2_low <= r2_high);
+
+  // Special case: No overlap, intervals are disjoint
+  if (r1_low > r2_high || r1_high < r2_low)
     return 0.0;
+  // Special case: All overlap, interval r2 is a subset of interval r1
+  if (r1_low <= r2_low && r2_high <= r1_high)
+    return 1.0;
+  /*
+   * At this point we know that r2_low < r2_high, because we would have returned
+   * by now otherwise. Note that for floating point types, however, we cannot
+   * conclude that r2_high - r2_low > 0 because of the possibility of underflow.
+   */
+  auto intersection_low = std::max(r1_low, r2_low);
+  auto intersection_high = std::min(r1_high, r2_high);
+  /*
+   * The intersection is a proper subset of interval r1, either the upper bounds
+   * are distinct, or lower bounds are distinct, or both. This means that any
+   * result has to be strictly greater than zero and strictly less than 1.
+   */
+  /*
+   * Guard against overflow. If the outer interval has a bound with an absolute
+   * value that's "too large", meaning that it might lead to an overflow, we
+   * apply a shrinking transformation that preserves the ratio. We only need to
+   * check the outer interval because the intersection is strictly smaller. For
+   * unsigned types we only need to check the upper bound, since the lower bound
+   * is zero. For signed types we need to check both.
+   */
+  const T safe_upper = std::nextafter(std::numeric_limits<T>::max() / 2, 0);
+  const T safe_lower = std::nextafter(std::numeric_limits<T>::min() / 2, 0);
+  bool unsafe = safe_upper < r2_high;
+  if (std::numeric_limits<T>::is_signed) {
+    unsafe = unsafe || r2_low < safe_lower;
+  }
+  if (unsafe) {
+    r2_low /= 2;
+    r2_high /= 2;
+    intersection_low /= 2;
+    intersection_high /= 2;
+  }
 
   // Compute ratio
-  auto overlap_start = std::max(d1[0], d2[0]);
-  auto overlap_end = std::min(d1[1], d2[1]);
-  auto overlap_range = overlap_end - overlap_start;
-  auto mbr_range = d2[1] - d2[0];
-  auto max = std::numeric_limits<double>::max();
+  auto numerator = intersection_high - intersection_low;
+  auto denominator = r2_high - r2_low;
   if (std::numeric_limits<T>::is_integer) {
-    overlap_range += 1;
-    mbr_range += 1;
+    // integer types count elements; they don't measure length
+    numerator += 1;
+    denominator += 1;
   } else {
-    if (overlap_range == 0)
-      overlap_range = std::nextafter(overlap_range, max);
-    if (mbr_range == 0)
-      mbr_range = std::nextafter(mbr_range, max);
+    if (denominator == 0) {
+      /*
+       * If the denominator is zero, we have no easy-to-access information about
+       * the quotient, but the result must still be between 0 and 1. The
+       * denominator is larger than the numerator, so the numerator will also be
+       * zero. We could extract some information from the floating-point
+       * representation itself, but that's a lot of work. So tra-la-la, this is
+       * a tiny defect.
```

Comment:
Nit: I'd take out the last sentence

Reply:
OK. I'll reword to be *more serious*. We do need to keep in an acknowledgement that it's a defect, albeit one that's quite difficult to reach.